tsdb/engine/bz1/bz1.go

```diff
@@ -0,0 +1,601 @@
+		}
+		return nil
+	})
+}
+
+// writeIndex writes a set of points for a single key.
+func (e *Engine) writeIndex(tx *bolt.Tx, key string, a [][]byte) error {
+	// Ignore if there are no points.
+	if len(a) == 0 {
+		return nil
+	}
+
+	// Create or retrieve series bucket.
+	bkt, err := tx.Bucket([]byte("points")).CreateBucketIfNotExists([]byte(key))
+	if err != nil {
+		return fmt.Errorf("create series bucket: %s", err)
+	}
+	c := bkt.Cursor()
+
+	// Ensure the slice is sorted before retrieving the time range.
+	sort.Sort(byteSlices(a))
+
+	// Determine time range of new data.
+	tmin, tmax := int64(btou64(a[0][0:8])), int64(btou64(a[len(a)-1][0:8]))
+
+	// If tmin is after the last block then append new blocks.
+	//
+	// This is the optimized fast path. Otherwise we need to merge the points
+	// with existing blocks on disk and rewrite all the blocks for that range.
+	if k, v := c.Last(); k == nil || int64(btou64(v[0:8])) < tmin {
+		if err := e.writeBlocks(bkt, a); err != nil {
+			return fmt.Errorf("append blocks: %s", err)
+		}
+	}
+
+	// Generate map of inserted keys.
+	m := make(map[int64]struct{})
+	for _, b := range a {
+		m[int64(btou64(b[0:8]))] = struct{}{}
+	}
+
+	// If time range overlaps existing blocks then unpack full range and reinsert.
+	for k, v := c.First(); k != nil; k, v = c.Next() {
+		// Determine block range.
+		bmin, bmax := int64(btou64(k)), int64(btou64(v[0:8]))
+
+		// Exit loop if the lowest block time is greater than the max insert time.
+		if bmin > tmax {
+			break
+		}
+
+		// If range overlaps with inserted range then read out data and delete block.
+		if (bmin >= tmin && bmin <= tmax) || (bmax >= tmin && bmax <= tmax) {
+			// Decode block.
+			buf, err := snappy.Decode(nil, v[8:])
+			if err != nil {
+				return fmt.Errorf("decode block: %s", err)
+			}
+
+			// Copy out any entries that aren't being overwritten.
+			for _, entry := range SplitEntries(buf) {
+				if _, ok := m[int64(btou64(entry[0:8]))]; !ok {
+					a = append(a, entry)
+				}
+			}
+
+			// Delete block in database.
+			c.Delete()
+		}
+	}
+
+	// Sort entries before rewriting.
+	sort.Sort(byteSlices(a))
+
+	// Rewrite points to new blocks.
+	if err := e.writeBlocks(bkt, a); err != nil {
+		return fmt.Errorf("rewrite blocks: %s", err)
+	}
+
+	return nil
+}
+
+// writeBlocks writes point data to the bucket in blocks.
+func (e *Engine) writeBlocks(bkt *bolt.Bucket, a [][]byte) error {
+	var block []byte
+
+	// Group points into blocks by size.
+	tmin, tmax := int64(math.MaxInt64), int64(math.MinInt64)
+	for i, p := range a {
+		// Update block time range.
+		timestamp := int64(btou64(p[0:8]))
+		if timestamp < tmin {
+			tmin = timestamp
+		}
+		if timestamp > tmax {
+			tmax = timestamp
+		}
+
+		// Append point to the end of the block.
+		block = append(block, p...)
+
+		// If the block is larger than the target block size or this is the
+		// last point then flush the block to the bucket.
+		if len(block) >= e.BlockSize || i == len(a)-1 {
```

Comment:
My thought is that the WAL flushes each partition when it can. And that the amount of data each series has will be variable. I can ensure that each will always have at least some minimum amount, but it could be quite a bit more if one series was filling up. A certain percentage of the series in a WAL partition have to be ready before it flushes. I also have logic in there that will force a series to flush if it gets to a certain ceiling size. I just want to make sure that if I send you a series with 2MB (the default ceiling) it'll get broken up into multiple blocks. But also that if I send you a bunch of series with 34k, they don't get broken into a 32k block and a 2k block. Instead that it would just put those together, while evenly splitting up the much larger ones



> On Aug 5, 2015, at 12:08 PM, Ben Johnson <notifications@github.com> wrote:
> 
> In tsdb/engine/bz1/bz1.go:
> 
> > +	for i, p := range a {
> > +		// Update block time range.
> > +		timestamp := int64(btou64(p[0:8]))
> > +		if timestamp < tmin {
> > +			tmin = timestamp
> > +		}
> > +		if timestamp > tmax {
> > +			tmax = timestamp
> > +		}
> > +
> > +		// Append point to the end of the block.
> > +		block = append(block, p...)
> > +
> > +		// If the block is larger than the target block size or this is the
> > +		// last point then flush the block to the bucket.
> > +		if len(block) >= e.BlockSize || i == len(a)-1 {
> It can't be a range since if you send the engine 3K of data it'll write a 3K block. Or if you send it 65K of data then it'll be a 64K block and a 1K block. Are you thinking it would reject writes that don't fit in a whole block back to the WAL? That seems complicated.
> 
> —
> Reply to this email directly or view it on GitHub.
>